File: android-core/src/androidTest/java/com/mparticle/integration_tests/IdentityApiOutgoingTest.java
package com.mparticle.integration_tests;

import com.mparticle.BaseCleanStartedEachTest;
import com.mparticle.MParticle;
import com.mparticle.identity.IdentityApiRequest;

import org.junit.Test;

import java.util.Random;

import static com.github.tomakehurst.wiremock.client.WireMock.matchingJsonPath;
import static com.github.tomakehurst.wiremock.client.WireMock.postRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.urlPathMatching;

public class IdentityApiOutgoingTest extends BaseCleanStartedEachTest {

    @Test
    public void testLogin() throws Exception {
        MParticle.getInstance().Identity().login();
        mServer.waitForVerify(postRequestedFor(urlPathMatching("/v([0-9]*)/login")).withRequestBody(matchingJsonPath(String.format("$.[?(@.previous_mpid == '%s')]", String.valueOf(mStartingMpid)))), 5000);
    }

    @Test
    public void testLoginNonEmpty() throws Exception {
        MParticle.getInstance().Identity().login(IdentityApiRequest.withEmptyUser().build());
        mServer.waitForVerify(postRequestedFor(urlPathMatching("/v([0-9]*)/login")).withRequestBody(matchingJsonPath(String.format("$.[?(@.previous_mpid == '%s')]", String.valueOf(mStartingMpid)))), 5000);
    }

    @Test
    public void testLogout() throws Exception {
        MParticle.getInstance().Identity().logout();
        mServer.waitForVerify(postRequestedFor(urlPathMatching("/v([0-9]*)/logout")).withRequestBody(matchingJsonPath(String.format("$.[?(@.previous_mpid == '%s')]", String.valueOf(mStartingMpid)))), 5000);
    }

    @Test
    public void testLogoutNonEmpty() throws Exception {
        MParticle.getInstance().Identity().logout(IdentityApiRequest.withEmptyUser().build());
        mServer.waitForVerify(postRequestedFor(urlPathMatching("/v([0-9]*)/logout")).withRequestBody(matchingJsonPath(String.format("$.[?(@.previous_mpid == '%s')]", String.valueOf(mStartingMpid)))), 5000);
    }

    @Test
    public void testModify() throws Exception {
        MParticle.getInstance().Identity().modify(IdentityApiRequest.withEmptyUser().build());
        mServer.waitForVerify(postRequestedFor(urlPathMatching(String.format("/v([0-9]*)/%s/modify", mStartingMpid))), 5000);
    }

    @Test
    public void testIdentify() throws Exception {
        MParticle.getInstance().Identity().identify(IdentityApiRequest.withEmptyUser().build());
        mServer.waitForVerify(postRequestedFor(urlPathMatching("/v([0-9]*)/identify")).withRequestBody(matchingJsonPath(String.format("$.[?(@.previous_mpid == '%s')]", String.valueOf(mStartingMpid)))), 5000);
    }


    @Override
    protected void beforeClass() throws Exception {

    }

    @Override
    protected void before() throws Exception {

    }
}

File: android-core/src/androidTest/kotlin/com.mparticle/identity/IdentityApiOutgoingTest.kt
package com.mparticle.identity

import com.mparticle.MParticle
import com.mparticle.networking.IdentityRequest.IdentityRequestBody
import com.mparticle.networking.Matcher
import com.mparticle.networking.MockServer.IdentityMatcher
import com.mparticle.testutils.BaseCleanStartedEachTest
import org.junit.Test

class IdentityApiOutgoingTest : BaseCleanStartedEachTest() {
    @Test
    @Throws(Exception::class)
    fun testLogin() {
        MParticle.getInstance()?.Identity()?.login()
        mServer.waitForVerify(Matcher(mServer.Endpoints().loginUrl).bodyMatch(object :
            IdentityMatcher() {
            override fun isIdentityMatch(identityRequest: IdentityRequestBody): Boolean {
                return mStartingMpid == identityRequest.previousMpid
            }
        }))
    }

    @Test
    @Throws(Exception::class)
    fun testLoginNonEmpty() {
        MParticle.getInstance()?.Identity()?.login(IdentityApiRequest.withEmptyUser().build())
        mServer.waitForVerify(Matcher(mServer.Endpoints().loginUrl).bodyMatch(object :
            IdentityMatcher() {
            override fun isIdentityMatch(identityRequest: IdentityRequestBody): Boolean {
                return mStartingMpid == identityRequest.previousMpid
            }
        }))
    }

    @Test
    @Throws(Exception::class)
    fun testLogout() {
        MParticle.getInstance()?.Identity()?.logout()
        mServer.waitForVerify(Matcher(mServer.Endpoints().logoutUrl).bodyMatch(object :
            IdentityMatcher() {
            override fun isIdentityMatch(identityRequest: IdentityRequestBody): Boolean {
                return mStartingMpid == identityRequest.previousMpid
            }
        }))
    }

    @Test
    @Throws(Exception::class)
    fun testLogoutNonEmpty() {
        MParticle.getInstance()?.Identity()?.logout(IdentityApiRequest.withEmptyUser().build())
        mServer.waitForVerify(Matcher(mServer.Endpoints().logoutUrl).bodyMatch(object :
            IdentityMatcher() {
            override fun isIdentityMatch(identityRequest: IdentityRequestBody): Boolean {
                return mStartingMpid == identityRequest.previousMpid
            }
        }))
    }

    @Test
    @Throws(Exception::class)
    fun testModify() {
        MParticle.getInstance()
            ?.Identity()?.modify(
                IdentityApiRequest.withEmptyUser().customerId(ran.nextLong().toString() + "")
                    .build()
            )
        mServer.waitForVerify(Matcher(mServer.Endpoints().getModifyUrl(mStartingMpid)))
    }

    @Test
    @Throws(Exception::class)
    fun testIdentify() {
        MParticle.getInstance()?.Identity()?.identify(IdentityApiRequest.withEmptyUser().build())
        mServer.waitForVerify(Matcher(mServer.Endpoints().identifyUrl).bodyMatch(object :
            IdentityMatcher() {
            override fun isIdentityMatch(identityRequest: IdentityRequestBody): Boolean {
                return mStartingMpid == identityRequest.previousMpid
            }
        }))
    }
}